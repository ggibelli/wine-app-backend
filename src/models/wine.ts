import mongoose, { Schema, Document } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { EspressioneComunitaria, DenomZona } from '../types';

export interface IWine {
  denominazioneVino: string;
  tipoVino?: string;
  espressioneComunitaria: EspressioneComunitaria;
  denominazioneZona: DenomZona;
  vitigni: mongoose.Types.Array<string>;
}

export interface WineGraphQl extends IWine {
  _id: mongoose.Types.ObjectId;
}

export interface IWineDoc extends IWine, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const wineSchemaFields: Record<keyof IWine, any> = {
  denominazioneVino: {
    type: String,
    required: true,
    minlength: 2,
    unique: true,
  },
  tipoVino: String,
  espressioneComunitaria: {
    type: String,
    enum: ['DOP', 'IGP', 'ND'],
  },
  denominazioneZona: {
    type: String,
    enum: ['DOC', 'DOCG', 'IGT', 'Vino varietale', 'Vino generico'],
  },
  vitigni: [
    {
      type: String,
    },
  ],
};

const wineSchema = new Schema(wineSchemaFields);

wineSchema.index({ denominazioneVino: 1 });

wineSchema.plugin(mongooseUniqueValidator);

export const Wine = mongoose.model<IWineDoc>('Wine', wineSchema);
