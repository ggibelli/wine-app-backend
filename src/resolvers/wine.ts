import { Resolvers, WineInput, WineInputUpdate } from '../generated/graphql';
import Users from '../data-sources/users';
import Wines from '../data-sources/wines';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  wines: Wines;
}

const resolver: StringIndexed<Resolvers> = {
  Query: {
    async wines(_, __, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.wines.getWines();
    },
    async wine(_, { id }, { dataSources }: { dataSources: MongoDataSource }) {
      return dataSources.wines.getWine(id);
    },
  },
  Mutation: {
    async createWine(
      _,
      { wine }: { wine: WineInput },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.createWine(wine);
    },

    async updateWine(
      _,
      { wine }: { wine: WineInputUpdate },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.updateWine(wine);
    },
    async deleteWine(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource },
    ) {
      return dataSources.wines.deleteWine(id);
    },
  },
};

export default resolver;
