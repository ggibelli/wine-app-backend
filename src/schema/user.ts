import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type ProducedWines {
    wine: Wine!
    bottlesProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  type OwnedVineyards {
    vineyard: Vineyard!
    tonsProduced: Int!
    metodoProduttivo: MetodoProduttivo!
  }

  input ProducedWinesInput {
    wine: ID!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input OwnedVineyardsInput {
    vineyard: ID!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input UserInput {
    email: String!
    password: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: AddressInput!
    hideContact: Boolean!
    #verified: Boolean
    #premium: Boolean
    #ads: [ID!]
    #negotiations: [ID!]
    #reviews: [ID!]
    #adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  input UserInputUpdate {
    _id: ID!
    email: String
    password: String
    firstName: String
    lastName: String
    pIva: String
    phoneNumber: String
    address: AddressInput
    isVerified: Boolean
    isPremium: Boolean
    hideContact: Boolean!
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  type Coordinates {
    latitude: Float
    longitude: Float
  }

  type User {
    _id: ID!
    email: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: Address!
    isVerified: Boolean!
    coordinates: Coordinates
    isPremium: Boolean
    isAdmin: Boolean!
    hideContact: Boolean!
    ads: [Ad!]
    savedAds: [Ad!]
    messages: [Message!]
    negotiations: [Negotiation!]
    reviews: [Review!]
    adsRemaining: Int @authorized
    dateCreated: Date! @date
    producedWines: ProducedWines
    ownedVineyards: OwnedVineyards
  }

  type AuthUser {
    token: String!
    user: User!
  }

  type AuthUserPayload {
    response: AuthUser
    errors: [Errors]
  }

  type UserPayload {
    response: User
    errors: [Errors]
  }

  extend type Query {
    users: [User!] @authenticated
    user(id: ID!): User @authenticated
    me: User @authenticated
  }

  extend type Mutation {
    createUser(user: UserInput!): AuthUserPayload
    updateUser(user: UserInputUpdate!): UserPayload @authenticated
    deleteUser(id: ID!): UserPayload @authenticated
    login(email: String!, password: String!): AuthUserPayload
  }
`;
