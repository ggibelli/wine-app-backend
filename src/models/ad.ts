/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
