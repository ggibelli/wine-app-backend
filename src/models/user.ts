/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
