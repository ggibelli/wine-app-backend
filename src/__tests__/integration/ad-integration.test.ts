/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../../tests/integrationSetup';
//import { ObjectId } from 'mongodb';
import { User } from '../../models/user';
import { Ad, AdGraphQl } from '../../models/ad';
import { users, ads } from '../../tests/mocksTests';
import {
  AdInput,
  AdInputUpdate,
  Wine as WineType,
} from '../../generated/graphql';
import {
  TypeAd,
  TypeProduct,
  Province,
  Regioni,
  MetodoProduttivo,
} from '../../types';
import createWineDb from '../../utils/wineExtractor';
import { Wine } from '../../models/wine';
//import { ObjectId } from 'mongodb';
import cron from 'cron';

const fakeStart = jest.fn(() => {
  fakeStop();
});
const fakeStop = jest.fn(() => null);
const fakeTick = jest.fn();

const FakeCron = jest.fn(() => ({
  stop: fakeStop,
  start: fakeStart,
  fireOnTick: fakeTick,
}));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
cron.CronJob = FakeCron;
const indirizzo = {
  via: 'asd asddasd',
  CAP: 12345,
  comune: 'aaaa',
  provincia: Province.AT,
  regione: Regioni.PIEMONTE,
};

const { query, mutate, setOptions } = testClient;

const ADS = gql`
  fragment AdPreview on Ad {
    datePosted
    postedBy {
      firstName
    }
    activeNegotiations
    harvest
    abv
    priceFrom
    priceTo
  }
  {
    ads(typeAd: BUY, typeProduct: AdWine) {
      ...AdPreview
    }
  }
`;

const AD = gql`
  query GetAd($id: ID!) {
    ad(id: $id) {
      content
      #datePosted
      numberViews
    }
  }
`;

const CREATE_AD = gql`
  mutation CreateAd($ad: AdInput!) {
    createAd(input: $ad) {
      response {
        content
        #datePosted
        postedBy {
          firstName
        }
        activeNegotiations
        harvest
        abv
        priceFrom
        priceTo
        ... on AdWine {
          wineName
          wine {
            regione
          }
        }
        ... on AdGrape {
          vineyardName
        }
      }
      errors {
        name
        text
      }
    }
  }
`;

const UPDATE_AD = gql`
  mutation UpdateAd($ad: AdInputUpdate!) {
    updateAd(input: $ad) {
      response {
        isActive
        content
        #datePosted
        postedBy {
          firstName
        }
      }
      errors {
        name
        text
      }
    }
  }
`;

const DELETE_AD = gql`
  mutation DeleteAd($id: ID!) {
    deleteAd(id: $id) {
      response {
        isActive
        content
        datePosted
        postedBy {
          firstName
        }
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID = gql`
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_OTHER = gql`
  mutation {
    login(email: "gio2@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

const LOGIN_VALID_NOT_VERIFIED = gql`
  mutation {
    login(email: "prova@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
        text
      }
    }
  }
`;

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const adsMock = ads();
  const wineDb = await createWineDb();
  const user = new User(usersMock[0]);
  const otherUser = new User(usersMock[1]);
  const thirdUser = new User(usersMock[2]);

  const ad = new Ad({ ...adsMock[0], postedBy: user });
  const wine = new Wine(wineDb[0]);
  await user.save();
  await otherUser.save();
  await thirdUser.save();
  await ad.save();
  await wine.save();
});

describe('Integration test ads', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('query ads wine in sale fails if not logged in', async () => {
    const res = await query(ADS);
    expect(res).toMatchSnapshot();
  });

  it('query single ad does not show 0 number visit if not logged', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await query(AD, {
      variables: { id: ad[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create adGrape mutation it fails if not logged in', async () => {
    const ad: AdInput = {
      vineyardName: 'Uvetta',
      metodoProduttivo: MetodoProduttivo.BIOLOGICO,
      harvest: 2020,
      abv: 12.0,
      priceFrom: 1.0,
      priceTo: 1.5,
      content: 'buona uva',
      address: indirizzo,
      typeAd: TypeAd.SELL,
      typeProduct: TypeProduct.ADGRAPE,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('create adWine mutation fails if user not verified', async () => {
    const data: any = await mutate(LOGIN_VALID_NOT_VERIFIED);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const wine: WineType[] = await Wine.find({}).lean().exec();
    const ad: AdInput = {
      wineName: wine[0].denominazioneVino,
      sottoZona: 'Sotto',
      harvest: 2020,
      abv: 12.0,
      priceFrom: 1.0,
      priceTo: 1.5,
      litersFrom: 500,
      litersTo: 600,
      content: 'wow',
      address: indirizzo,
      typeAd: TypeAd.BUY,
      typeProduct: TypeProduct.ADWINE,
      needsFollowUp: true,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('query ads wine in sale succeds if logged in', async () => {
    const data: any = await mutate(LOGIN_VALID);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token: string = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const res = await query(ADS);
    expect(res).toMatchSnapshot();
  });

  it('create adWine mutation successfull', async () => {
    const wine: WineType[] = await Wine.find({}).lean().exec();
    const ad: AdInput = {
      wineName: wine[0].denominazioneVino,
      sottoZona: 'Sotto',
      harvest: 2020,
      abv: 12.0,
      priceFrom: 1.0,
      priceTo: 1.5,
      litersFrom: 500,
      litersTo: 600,
      content: 'wow',
      address: indirizzo,
      typeAd: TypeAd.BUY,
      typeProduct: TypeProduct.ADWINE,
      needsFollowUp: true,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('update adWine mutation succeds if logged in and same user', async () => {
    const adToUpdate: AdGraphQl[] = await Ad.find({}).lean().exec();
    const ad: AdInputUpdate = {
      _id: adToUpdate[0]._id.toString(),
      isActive: false,
    };
    const res = await mutate(UPDATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete adWine mutation succeds if logged in and same user', async () => {
    const adToDelete: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await mutate(DELETE_AD, {
      variables: { id: adToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('update adWine mutation fails if logged in and other user', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const adToUpdate: AdGraphQl[] = await Ad.find({}).lean().exec();
    const ad: AdInputUpdate = {
      _id: adToUpdate[0]._id.toString(),
      isActive: false,
    };
    const res = await mutate(UPDATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('delete adWine mutation fails if logged in and other user', async () => {
    const data: any = await mutate(LOGIN_VALID_OTHER);
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    const token = data.data.login.response.token;
    setOptions({
      request: {
        headers: {
          authorization: token,
        },
      },
    });
    const adToDelete: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await mutate(DELETE_AD, {
      variables: { id: adToDelete[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });

  it('create adWine mutation successfull, followup sent to user looking to buy wine', async () => {
    const wine: WineType[] = await Wine.find({}).lean().exec();
    const ad: AdInput = {
      wineName: wine[0].denominazioneVino,
      harvest: 2020,
      abv: 12.5,
      priceFrom: 1.0,
      priceTo: 1.5,
      litersFrom: 500,
      litersTo: 600,
      content: 'wow',
      address: indirizzo,
      typeAd: TypeAd.SELL,
      typeProduct: TypeProduct.ADWINE,
      needsFollowUp: false,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
    expect(fakeStart).toBeCalledTimes(1);
    expect(fakeStop).toBeCalledTimes(1);
  });

  it('create adGrape mutation successfull', async () => {
    const ad: AdInput = {
      vineyardName: 'Uvetta',
      metodoProduttivo: MetodoProduttivo.BIOLOGICO,
      harvest: 2020,
      abv: 12.0,
      priceFrom: 1.0,
      priceTo: 1.5,
      kgFrom: 500,
      kgTo: 600,
      content: 'buona uva',
      address: indirizzo,
      typeAd: TypeAd.SELL,
      typeProduct: TypeProduct.ADGRAPE,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('create adGrape mutation it fails if kgs are missing', async () => {
    const ad: AdInput = {
      vineyardName: 'Uvetta',
      metodoProduttivo: MetodoProduttivo.BIOLOGICO,

      harvest: 2020,
      abv: 12.0,
      priceFrom: 1.0,
      priceTo: 1.5,
      //kgFrom: 500,
      //kgTo: 600,
      content: 'buona uva',
      address: indirizzo,
      typeAd: TypeAd.SELL,
      typeProduct: TypeProduct.ADGRAPE,
    };
    const res = await mutate(CREATE_AD, {
      variables: { ad },
    });
    expect(res).toMatchSnapshot();
  });

  it('query single ad shows number visit if logged', async () => {
    const ad: AdGraphQl[] = await Ad.find({}).lean().exec();
    const res = await query(AD, {
      variables: { id: ad[0]._id.toString() },
    });
    expect(res).toMatchSnapshot();
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
