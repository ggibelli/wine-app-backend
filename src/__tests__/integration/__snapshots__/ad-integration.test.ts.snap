// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test ads create adGrape mutation it fails if kgs are missing 1`] = `
Object {
  "data": Object {
    "createAd": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
          "text": "The kg from is mandatory",
        },
        Object {
          "name": "UserInputError",
          "text": "The kg to is mandatory",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test ads create adGrape mutation it fails if not logged in 1`] = `
Object {
  "data": Object {
    "createAd": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "createAd",
      ],
    },
  ],
}
`;

exports[`Integration test ads create adGrape mutation successfull 1`] = `
Object {
  "data": Object {
    "createAd": Object {
      "errors": Array [],
      "response": Object {
        "abv": 12,
        "activeNegotiations": 0,
        "content": "buona uva",
        "harvest": 2020,
        "postedBy": Object {
          "firstName": "Mariuccio",
        },
        "priceFrom": 1,
        "priceTo": 1.5,
        "vineyardName": "Uvetta",
      },
    },
  },
}
`;

exports[`Integration test ads create adWine mutation fails if user not verified 1`] = `
Object {
  "data": Object {
    "createAd": Object {
      "errors": Array [
        Object {
          "name": "AuthorizationError",
          "text": "You need to verify your account",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test ads create adWine mutation successfull 1`] = `
Object {
  "data": Object {
    "createAd": Object {
      "errors": Array [],
      "response": Object {
        "abv": 12,
        "activeNegotiations": 0,
        "content": "wow",
        "harvest": 2020,
        "postedBy": Object {
          "firstName": "Giovanni",
        },
        "priceFrom": 1,
        "priceTo": 1.5,
        "wineName": "Abruzzo",
      },
    },
  },
}
`;

exports[`Integration test ads create adWine mutation successfull, followup sent to user looking to buy wine 1`] = `
Object {
  "data": Object {
    "createAd": Object {
      "errors": Array [],
      "response": Object {
        "abv": 12.5,
        "activeNegotiations": 0,
        "content": "wow",
        "harvest": 2020,
        "postedBy": Object {
          "firstName": "Mariuccio",
        },
        "priceFrom": 1,
        "priceTo": 1.5,
        "wineName": "Abruzzo",
      },
    },
  },
}
`;

exports[`Integration test ads delete adWine mutation fails if logged in and other user 1`] = `
Object {
  "data": Object {
    "deleteAd": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Error during the delete, only ad owner can delete it",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test ads delete adWine mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "deleteAd": Object {
      "errors": Array [],
      "response": Object {
        "content": "wow",
        "datePosted": "17 Dec 95, 3:24",
        "isActive": false,
        "postedBy": Object {
          "firstName": "Giovanni",
        },
      },
    },
  },
}
`;

exports[`Integration test ads query ads wine in sale fails if not logged in 1`] = `
Object {
  "data": Object {
    "ads": Object {
      "ads": Array [
        null,
      ],
    },
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 7,
          "line": 5,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "ads",
        "ads",
        0,
        "postedBy",
      ],
    },
  ],
}
`;

exports[`Integration test ads query ads wine in sale succeds if logged in 1`] = `
Object {
  "data": Object {
    "ads": Object {
      "ads": Array [
        Object {
          "abv": 12,
          "activeNegotiations": 0,
          "datePosted": "17 Dec 95, 3:24",
          "harvest": 2020,
          "postedBy": Object {
            "firstName": "Giovanni",
          },
          "priceFrom": 1,
          "priceTo": 1.5,
        },
      ],
    },
  },
}
`;

exports[`Integration test ads query single ad does not show 0 number visit if not logged 1`] = `
Object {
  "data": Object {
    "ad": Object {
      "content": "wow",
      "numberViews": 0,
    },
  },
}
`;

exports[`Integration test ads query single ad shows number visit if logged 1`] = `
Object {
  "data": Object {
    "ad": Object {
      "content": "wow",
      "numberViews": 1,
    },
  },
}
`;

exports[`Integration test ads saveAd mutation it fails if not logged in 1`] = `
Object {
  "data": Object {
    "saveAd": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "UNAUTHENTICATED",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "You need to login to continue",
      "path": Array [
        "saveAd",
      ],
    },
  ],
}
`;

exports[`Integration test ads saveAd mutation succeds if logged in and if ad is already saved it gets removed 1`] = `
Object {
  "data": Object {
    "saveAd": Object {
      "errors": Array [],
      "response": Object {
        "content": "wow",
      },
    },
  },
}
`;

exports[`Integration test ads saveAd mutation succeds if logged in and savedAd shows up in user savedads 1`] = `
Object {
  "data": Object {
    "saveAd": Object {
      "errors": Array [],
      "response": Object {
        "content": "wow",
      },
    },
  },
}
`;

exports[`Integration test ads update adWine mutation fails if logged in and other user 1`] = `
Object {
  "data": Object {
    "updateAd": Object {
      "errors": Array [
        Object {
          "name": "General error",
          "text": "Error during the update, only ad owner can update it",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test ads update adWine mutation succeds if logged in and same user 1`] = `
Object {
  "data": Object {
    "updateAd": Object {
      "errors": Array [],
      "response": Object {
        "content": "wow",
        "isActive": false,
        "postedBy": Object {
          "firstName": "Giovanni",
        },
      },
    },
  },
}
`;
